Framework/NexusGeometry/src/NexusGeometrySave.cpp

```diff
@@ -0,0 +1,806 @@
+ *
+ * @param compInfo : componentInfo object.
+ * @param fullPath : save destination as full path.
+ * @param reporter : report to progressBase.
+ */
+void saveInstrument(
+    const std::pair<std::unique_ptr<Geometry::ComponentInfo>,
+                    std::unique_ptr<Geometry::DetectorInfo>> &instrPair,
+    const std::string &fullPath, const std::string &rootPath,
+    Kernel::ProgressBase *reporter) {
+
+  const Geometry::ComponentInfo &compInfo = (*instrPair.first);
+  const Geometry::DetectorInfo &detInfo = (*instrPair.second);
+
+  // Exception handling.
+  boost::filesystem::path tmp(fullPath);
+  if (!boost::filesystem::is_directory(tmp.root_directory())) {
+    throw std::invalid_argument(
+        "The path provided for saving the file is invalid: " + fullPath + "\n");
+  }
+
+  // check the file extension matches any of the valid extensions defined in
+  // nexus_geometry_extensions
+  const auto ext = boost::filesystem::path(tmp).extension();
+  bool isValidExt = std::any_of(
+      nexus_geometry_extensions.begin(), nexus_geometry_extensions.end(),
+      [&ext](const std::string &str) { return ext.generic_string() == str; });
+
+  // throw if the file extension is invalid
+  if (!isValidExt) {
+
+    // string of valid extensions to output in exception
+    std::string extensions = "";
+    std::for_each(
+        nexus_geometry_extensions.begin(), nexus_geometry_extensions.end(),
+        [&extensions](const std::string &s) { extensions += " " + s; });
+    throw std::invalid_argument("invalid extension for file: '" +
+                                ext.generic_string() +
+                                "'. Expected any of: " + extensions);
+  }
+
+  if (!compInfo.hasDetectorInfo()) {
+    throw std::invalid_argument(
+        "No detector info was found in the Instrument cache.\n");
+  }
+  if (!compInfo.hasSample()) {
+    throw std::invalid_argument(
+        "No sample was found in the Instrument cache.\n");
+  }
+
+  if (Mantid::Kernel::V3D{0, 0, 0} != compInfo.samplePosition()) {
+    throw std::invalid_argument(
+        "The sample posiiton is required to be at the origin.\n");
+  }
+
+  if (!compInfo.hasSource()) {
+    throw std::invalid_argument("No source was found in the Instrument cache.");
+  }
+
+  const auto detIds = detInfo.detectorIDs();
+
+  // open file
+  H5::H5File file(fullPath, H5F_ACC_TRUNC); // open file
+
+  H5::Group rootGroup, instrument;
+
+  // create NXentry (file root)
+  rootGroup = file.createGroup(rootPath);
+  NexusGeometrySave::writeStrAttribute(rootGroup, NX_CLASS, NX_ENTRY);
+
+  // save and capture NXinstrument (component root)
+  instrument = NexusGeometrySave::NXInstrument(rootGroup, compInfo);
+
+  // save NXsource
+  NexusGeometrySave::saveNXSource(instrument, compInfo);
+
+  // save NXsample
+  NexusGeometrySave::saveNXSample(rootGroup, compInfo);
+
+  // save NXdetectors
+  for (size_t index = compInfo.root() - 1; index > 0; --index) {
```

Comment:
`index > compInfo.root() - compInfo.size()` - There is no need to loop over detectors.

Reply:
this doesnt do as expected: behaviour was detectors werent found in saveIstrument and tests fail. on 783, i added a break if isDetector. otherwise i am unsure

Reply:
done